Lineage model fitting - sketchlib changes

nickjcroucher opened the pull request:
- Needed for https://github.com/bacpop/PopPUNK/compare/master...parsing_fixes?expand=1
- This should make the kNN matrices returned by CPU and GPU consistent, both with predictable memory use
- One obvious inefficiency is copying `combine_vectors` into `matrix_ops.cpp` from PopPUNK for CPU multithreading - I don't know if the PopPUNK version could be deleted and the function referenced from sketchlib instead
- Will update documentation if changes are satisfactory - validation comes from the PopPUNK output in linked PR above

johnlees commented on pp_sketch/__main__.py:
```diff
@@ -145,6 +145,14 @@ def get_options():
         else:
             arguments['--kNN'] = 0
             arguments['--threshold'] = float(arguments['--threshold'])
+        if arguments['--require-reciprocity']:
+            arguments['--require-reciprocity'] = True
+        else:
+            arguments['--require-reciprocity'] = False
+        if arguments['--count-all-neighbours']:
+            arguments['--count-all-neighbours'] = True
+        else:
+            arguments['--count-all-neighbours'] = False
```
You don't need this, you can just use `if arguments['--require-reciprocity']` later

nickjcroucher replied:
Gone.

johnlees commented on src/dist/matrix_ops.cpp:
```diff
@@ -25,55 +23,89 @@ void rectangle_block(const Eigen::VectorXf &longDists,
+                          const float distCutoff,
+                          const unsigned long int kNN,
+                          const size_t num_threads)
+{
+    if (kNN > 0 && distCutoff > 0)
+    {
+        throw std::runtime_error("Specify only one of kNN or distCutoff");
+    }
+    else if (kNN < 1 && distCutoff < 0)
+    {
+        throw std::runtime_error("kNN must be > 1 or distCutoff > 0");
```
Brace formatting is inconsistent here. I usually use `clang-format` (you can add this into vscode to run on C++ files)

nickjcroucher replied:
Done

johnlees commented on src/dist/matrix_ops.cpp:
```diff
@@ -25,55 +23,89 @@ void rectangle_block(const Eigen::VectorXf &longDists,
+                {
+                    continue; // Ignore diagonal which will always be one of the closest
+                }
+                if (counter < kNN)
```
I think you want the `new_val` check, as in the old code? Otherwise this always adds exactly `k`, and you can use the more easily parallelisable version

nickjcroucher replied:
Deleted

johnlees commented on src/sketchlib_bindings.cpp:
```diff
@@ -217,8 +218,14 @@ double jaccardDist(const std::string &db_name, const std::string &sample1,
+                         const float distCutoff,
+                         const unsigned long int kNN,
+                         const size_t num_threads)
+{
```
Braces here too – run clang-format on all updated files

nickjcroucher replied:
Done

johnlees reviewed (changes requested):
Two larger things:
- I don't actually understand what is being attempted here: reciprocity and all neighbour options have been added, but not used. The C++ code is changed to add parallelisation, but also so that exactly k matches are reported in kNN modes. What are you actually trying to change?
- It would make sense to have a check of the results which is independent of PopPUNK, and not just a check that it all runs. This should be easy to add with some test data.

> One obvious inefficiency is copying combine_vectors into matrix_ops.cpp from PopPUNK for CPU multithreading - I don't know if the PopPUNK version could be deleted and the function referenced from sketchlib instead

Yes, we should just have one function. We can check this when we do https://github.com/bacpop/PopPUNK/pull/232.

> Will update documentation if changes are satisfactory - validation comes from the PopPUNK output in linked PR above

Should just need a small change to the README

nickjcroucher commented:
The overall changes are described better in the associated PopPUNK PR: https://github.com/bacpop/PopPUNK/pull/232 - apologies for leaving the old changes in the docopt string, they pre-date the alterations to the GPU kNN code.

nickjcroucher commented:
So the `sparsify_dists` function is not actually used for kNN clustering any more (I think it was when I started the branch) - instead it is done by `query_db_sparse` ([here](https://github.com/bacpop/pp-sketchlib/blob/a7ea4cce1a3b6b0e06fddf45e4b8551fc7c6382c/src/api.cpp#L314)), which is parallelised as you suggest already. The kNN mode of `sparsify_dists` was only used in `test-dists.py` and `test-matrix.py`, which were essentially validating unused bits of code. I've now modified these, such that `test-dists.py` compares against independent numpy functions - I've got rid of the unused code, and I've renamed the thresholding function to avoid any future misunderstandings.

nickjcroucher commented:
@muppi1993 identified that the kNN option of `sparsify_dists` is used by PopPUNK for lineage model fitting - hence the code has now been moved to `extend.cpp` in PopPUNK in https://github.com/bacpop/PopPUNK/commit/f0526c15acd0d0d7d6d8d0f0f94735c455074e03, as it is only used by PopPUNK, not by sketchlib.

johnlees commented on src/sketchlib_bindings.cpp:
```diff
@@ -137,12 +137,13 @@ NumpyMatrix queryDatabase(const std::string &ref_db_name,
-            const float dist_cutoff = 0.0f, const size_t dist_col = 0,
-            const size_t num_threads = 1, const bool use_gpu = false,
-            const int device_id = 0) {
+querySelfSparse(const std::string &ref_db_name,
+                const std::vector<std::string> &ref_names,
+                std::vector<size_t> kmer_lengths,
+                const bool random_correct = true, const bool jaccard = false,
+                const unsigned long int kNN = 0, const float dist_cutoff = 0.0f,
+                const size_t dist_col = 0, const size_t num_threads = 1,
+                const bool use_gpu = false, const int device_id = 0) {
```
What has changed here? Just formatting?

nickjcroucher replied:
Yep

johnlees reviewed (commented):
I think this all looks fine now

johnlees commented:
Also we need to bear in mind that mandrake also uses sketchlib, so any API changes should be checked there too (currently `queryDatabaseSparse` and `querySelfSparse`, so this PR should be fine)

nickjcroucher commented:
> I think this all looks fine now. I think it makes sense that sketchlib will just do sparse or dense queries, and that sparsifying dense results ends up elsewhere In PopPUNK.models, the lineage mode extend function is where I thought the sparsify function is called from. Is this all in the PopPUNK C++ extension now?

As a quick audit, there is similar code involved in:
- sparsification of distances in the fly in sketchlib (https://github.com/bacpop/pp-sketchlib/blob/413f57e2a392e90ac42fee846f2a6732a1073574/src/api.cpp#L314)
- fitting the lineage model in PopPUNK (https://github.com/bacpop/PopPUNK/blob/00ed0756409e3cfd991dfc408dc55a5e3ee8aead/src/extend.cpp#L237)
- extending the lineage model in PopPUNK (https://github.com/bacpop/PopPUNK/blob/00ed0756409e3cfd991dfc408dc55a5e3ee8aead/src/extend.cpp#L53)

johnlees commented:
@nickjcroucher do you want this version to be released, put on conda etc?

nickjcroucher commented:
Yes please, that will make the CI easier for the PopPUNK branch
________________________________
From: John Lees ***@***.***>
Sent: Monday, October 24, 2022 12:15:32 PM
To: bacpop/pp-sketchlib ***@***.***>
Cc: Croucher, Nicholas J ***@***.***>; Mention ***@***.***>
Subject: Re: [bacpop/pp-sketchlib] Lineage model fitting - sketchlib changes (PR #82)


@nickjcroucher<https://github.com/nickjcroucher> do you want this version to be released, put on conda etc?

—
Reply to this email directly, view it on GitHub<https://github.com/bacpop/pp-sketchlib/pull/82#issuecomment-1288879838>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD6PWQ6N7D4EA7PPVQCSEYTWEZVVJANCNFSM6AAAAAARGHKNQI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
